infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/receive/StreamsSnapshotWriter.java

```diff
@@ -0,0 +1,273 @@
+package org.corfudb.infrastructure.logreplication.receive;
+
+import io.netty.buffer.Unpooled;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.logreplication.LogReplicationConfig;
+import org.corfudb.protocols.logprotocol.OpaqueEntry;
+import org.corfudb.protocols.logprotocol.SMREntry;
+import org.corfudb.protocols.wireprotocol.logreplication.LogReplicationEntry;
+import org.corfudb.protocols.wireprotocol.logreplication.LogReplicationEntryMetadata;
+import org.corfudb.protocols.wireprotocol.logreplication.MessageType;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.CorfuStoreMetadata;
+import org.corfudb.runtime.collections.TxBuilder;
+import org.corfudb.runtime.view.StreamOptions;
+import org.corfudb.runtime.view.stream.OpaqueStream;
+import org.corfudb.util.serializer.Serializers;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.lang.reflect.Array;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Stream;
+
+/**
+ * Writing a snapshot fullsync data
+ * Open streams interested and append all entries
+ */
+
+@Slf4j
+@NotThreadSafe
+public class StreamsSnapshotWriter implements SnapshotWriter {
+    final static String SHADOW_STREAM_NAME_SUFFIX = "_shadow";
+    HashMap<UUID, String> streamViewMap; // It contains all the streams registered for write to.
+    HashMap<UUID, String> shadowMap;
+    CorfuRuntime rt;
+
+    long siteConfigID;
+    private long srcGlobalSnapshot; // The source snapshot timestamp
+    private long recvSeq;
+    private long shadowStreamStartAddress;
+    @Getter
+    private PersistedWriterMetadata persistedWriterMetadata;
+    HashMap<UUID, UUID> uuidMap;
+    Phase phase;
+
+
+    // The sequence number of the message, it has received.
+    // It is expecting the message in order of the sequence.
+
+    public StreamsSnapshotWriter(CorfuRuntime rt, LogReplicationConfig config, PersistedWriterMetadata persistedWriterMetadata) {
+        this.rt = rt;
+        this.persistedWriterMetadata = persistedWriterMetadata;
+        streamViewMap = new HashMap<>();
+        uuidMap = new HashMap<>();
+        shadowMap = new HashMap<>();
+        phase = Phase.TransferPhase;
```

Comment:
when will ApplyPhase be switched back to TransferPhase ? I do not see that. So after the first snapshot sync, how does it work the second time.